docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,437 @@
+          encapsulation: None
+          natOutgoing: Enabled
+          nodeSelector: all(){{end}}
+```
+
+### Additional considerations
+
+#### HelmChartProxy design
+
+The scope of the cluster values accessible using Go templating must be defined along with use cases. An option would be to expose the Builtin struct type used in ClusterClass or construct a similar type. Additionally, while resources like the control plane and cluster have a 1:1 correlation with each selected cluster, resources like MachineDeployments or Machines could have any number and can’t be selected in a similar way. As a result, additional work is needed to decide if there’s a use case for including information in those types.
+
+#### Provider contracts
+
+As it stands, there is no formalized provider contract between ClusterAddonProviders that a ClusterAddonProvider must implement. That being said, the design of ClusterAddonProvider for Helm does suggest that certain patterns and components can be reused between different ClusterAddonProviders. These patterns can include:
+
+- Designing a CRD like HelmChartProxy to specify an add-on to install on certain Clusters.
+- Using a `clusterSelector` to install an add-on on a selected workload Cluster and to uninstall an add-on a Cluster that is unselected.
+- Using Go templates in a field like `valuesTemplate` to dynamically resolve configurations based on each Cluster's definition.
+- Designing a CRD like HelmReleaseProxy to:
+  - Maintain an inventory of add-ons and differentiate between managed add-ons and out of band add-ons.
+  - Provide a modular design such that the HelmChartProxy controller does not need to implement any Helm operations and only make Kubernetes client calls to create, update, or delete a HelmReleaseProxy.
+
+## Cluster API enhancements
+
+Despite the fact that most of the heavy lifting for add-on management is delegated to external ClusterAddonProvider components, over time we would like to provide Cluster API’s users an integrated and seamless user experience for add-on orchestration.
+
+### clusterctl
+
+ClusterAddonProvider is a new component to be installed in a Management Cluster, and in order to make this operation simpler, we can extend clusterctl in order to manage ClusterAddonProvider implementations in clusterctl init, clusterctl upgrade, etc.
+
+Note: clusterctl assumes all the providers to be linked to a CAPI contract; while it is not clear at this stage if the same applies to ClusterAddonProvider. This should be further investigated.
+
+## Alternatives
+
+### Why not ClusterResourceSets?
+
+This proposal introduces ClusterAddonProviders as opposed to overhauling ClusterResourceSets for the following reasons:
+
+- ClusterResourceSets violate the Cluster API design philosophy of defining a desired state through a declarative spec and having controllers create, delete, or update resources to build the desired state. Instead, ClusterrResourceSets are designed to "apply once and forget."
```

Comment:
I disagree.

Community members [proposed](https://github.com/kubernetes-sigs/cluster-api/pull/6188) to extend CRS with an "apply always" strategy, i.e. to update when the desired resources have changed. This proposal is implemented in a Cluster API fork, and used in the production today.

Reply:
Thanks for sharing that proposal, I wasn't aware of that. Is it similar to the idea of #6555 to add a Reconcile() mode? I suppose we could afford to reword this section a bit. The main point I was trying to get across is that in the main branch of CAPI, CRS is a bit of an outlier in that it doesn't yet have a canonical Reconcile() mode and I don't believe there's a Delete() mode implemented either.